File: Data Structures & Algorithms/combination-target-sum/submission-1.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        n = len(nums)
        res = []

        def util(i, t, r):
            if t == 0:
                res.append(r.copy())
                return
            if t < 0 or i == n:
                return
            r.append(nums[i])
            util(i, t-nums[i], r)
            r.pop()
            util(i+1, t, r)
        
        util(0, target, [])
        return res

File: Data Structures & Algorithms/target-sum/submission-1.cpp
class Solution {
public:
    
    int util (vector<int>& nums, int target, int i, unordered_map<long long, int>& dp) {
        auto key = [](int a, int b) { return (long long)a << 32 | (unsigned)b; };
        if (i == nums.size() and target == 0) return 1;
        if (i == nums.size()) return 0;
        if (dp.find(key(i, target)) != dp.end()) return dp[key(i, target)];
        return dp[key(i, target)] = dp[key(i, target)]+(util(nums, target - nums[i], i+1, dp) + util(nums, target + nums[i], i+1, dp));
    }
    
    int findTargetSumWays(vector<int>& nums, int target) {
        unordered_map<long long, int> dp;
        return util(nums, target, 0, dp);
    }
};
